File: python/p034.py
# Digit factorials
#
# Find the sum of all numbers which are equal to the sum of the factorial of
# their digits.
from math import factorial
f = [ factorial(i) for i in range(10) ]
fsum = lambda n: sum(f[int(i)] for i in str(n))
result = sum( n for n in range(10,7*factorial(9)) if fsum(n) == n )
print(result)
# output: 40730

File: go/p034.go
package main

import "fmt"

func main() {
	sum := 0
	for i := 10; i < 7*fact(9); i++ {
		if factsum(i) == i {
			sum += i
		}
	}
	fmt.Println(sum)
	// output: 40730
}

func fact(n int) int {
	if n < 2 {
		return 1
	}
	return n * fact(n-1)
}

func factsum(n int) int {
	sum := 0
	for n > 0 {
		sum += fact(n % 10)
		n /= 10
	}
	return sum
}
